pkg/controller/deployment/deployment_controller.go

```diff
@@ -507,6 +496,51 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
 	return nil
 }
 
+func (dc *DeploymentController) getNewRCAndOldRCs(deployment extensions.Deployment) (*api.ReplicationController, []*api.ReplicationController, error) {
+	oldRCs, err := dc.getOldRCs(deployment)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	maxOldV, err := maxRevision(oldRCs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newRC, err := dc.getNewRC(deployment, maxOldV)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if newRC.Annotations != nil && newRC.Annotations[deploymentutil.RevisionAnnotation] == "" &&
+		deployment.Annotations == nil || deployment.Annotations[deploymentutil.RevisionAnnotation] != newRC.Annotations[deploymentutil.RevisionAnnotation] {
+		if err = dc.updateDeploymentRevision(deployment, newRC.Annotations[deploymentutil.RevisionAnnotation]); err != nil {
+			return nil, nil, err
+		}
+	}
+	return newRC, oldRCs, nil
+}
+
+func revision(rc *api.ReplicationController) (int, error) {
+	v, ok := rc.Annotations[deploymentutil.RevisionAnnotation]
+	if !ok {
+		return 0, nil
+	}
+	return strconv.Atoi(v)
+}
+
+func maxRevision(allRCs []*api.ReplicationController) (int, error) {
+	max := 0
+	for _, rc := range allRCs {
+		if v, err := revision(rc); err != nil {
```

Comment:
I don't think we should bail here in the case that one RC isn't properly annotated. We should degrade gracefully -- just skip it. It won't appear in the revision history in kubectl and perhaps it wouldn't be possible to automatically roll back to that revision, but all the other behavior should work correctly.